
--- a/openmp/cmake/modules/LibompHandleFlags.cmake	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/cmake/modules/LibompHandleFlags.cmake	2026-07-08 23:57:59.241585185 +0800
@@ -48,7 +48,7 @@
   libomp_append(flags_local /GS LIBOMP_HAVE_GS_FLAG)
   libomp_append(flags_local /EHsc LIBOMP_HAVE_EHSC_FLAG)
   libomp_append(flags_local /Oy- LIBOMP_HAVE_OY__FLAG)
-  if(${IA32} OR ${INTEL64})
+  if(${IA32} OR ${INTEL64} OR ${X32})
     libomp_append(flags_local -mrtm LIBOMP_HAVE_MRTM_FLAG)
   endif()
   # Intel(R) C Compiler flags
@@ -60,12 +60,16 @@
     libomp_append(flags_local -Qinline-min-size=1 LIBOMP_HAVE_INLINE_MIN_SIZE_FLAG)
   endif()
   # Architectural C and C++ flags
-  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+  if(${X32})
+    libomp_append(flags_local -mx32 LIBOMP_HAVE_M32_FLAG)
+  elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
     libomp_append(flags_local -m32 LIBOMP_HAVE_M32_FLAG)
   endif()
   if(${IA32})
     libomp_append(flags_local /arch:SSE2 LIBOMP_HAVE_ARCH_SSE2_FLAG)
     libomp_append(flags_local -msse2 LIBOMP_HAVE_MSSE2_FLAG)
+  elseif(${X32})
+    libomp_append(flags_local -msse2 LIBOMP_HAVE_MSSE2_FLAG)
     libomp_append(flags_local -falign-stack=maintain-16-byte LIBOMP_HAVE_FALIGN_STACK_FLAG)
   elseif(${MIC})
     libomp_append(flags_local -mmic LIBOMP_HAVE_MMIC_FLAG)
@@ -81,7 +85,9 @@
 function(libomp_get_asmflags asmflags)
   set(asmflags_local)
   # Architectural assembler flags
-  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+  if(${X32})
+    libomp_append(asmflags_local -mx32 LIBOMP_HAVE_M32_FLAG)
+  elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
     libomp_append(asmflags_local -m32 LIBOMP_HAVE_M32_FLAG)
   endif()
   if(${IA32})
@@ -117,7 +123,9 @@
   libomp_append(ldflags_local -static-intel LIBOMP_HAVE_STATIC_INTEL_FLAG)
   libomp_append(ldflags_local /SAFESEH LIBOMP_HAVE_SAFESEH_FLAG)
   # Architectural linker flags
-  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+  if(${X32})
+    libomp_append(ldflags_local -mx32 LIBOMP_HAVE_M32_FLAG)
+  elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
     libomp_append(ldflags_local -m32 LIBOMP_HAVE_M32_FLAG)
   endif()
   if(${IA32})
@@ -164,7 +172,9 @@
 # Fortran flags
 function(libomp_get_fflags fflags)
   set(fflags_local)
-  if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+  if(${X32})
+    libomp_append(fflags_local -mx32 LIBOMP_HAVE_M32_FORTRAN_FLAG)
+  elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
     libomp_append(fflags_local -m32 LIBOMP_HAVE_M32_FORTRAN_FLAG)
   endif()
   set(fflags_local ${fflags_local} ${LIBOMP_FFLAGS})
@@ -181,6 +191,9 @@
   elseif(${INTEL64})
     set(libomp_gdflag_arch arch_32e)
     libomp_append(gdflags_local "-D IS_IA_ARCH")
+  elseif(${X32})
+    set(libomp_gdflag_arch arch_x32)
+    libomp_append(gdflags_local "-D IS_IA_ARCH")
   else()
     set(libomp_gdflag_arch arch_${LIBOMP_ARCH})
   endif()
--- a/openmp/cmake/modules/LibompUtils.cmake	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/cmake/modules/LibompUtils.cmake	2026-07-08 23:57:59.241860287 +0800
@@ -91,6 +91,8 @@
     set(${return_arch_string} "IA-32" PARENT_SCOPE)
   elseif(${INTEL64})
     set(${return_arch_string} "Intel(R) 64" PARENT_SCOPE)
+  elseif(${X32})
+    set(${return_arch_string} "x32" PARENT_SCOPE)
   elseif(${MIC})
     set(${return_arch_string} "Intel(R) Many Integrated Core Architecture" PARENT_SCOPE)
   elseif(${ARM})
--- a/openmp/runtime/CMakeLists.txt	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/CMakeLists.txt	2026-07-08 23:57:59.242101621 +0800
@@ -40,9 +40,17 @@
   elseif(LIBOMP_NATIVE_ARCH STREQUAL "x86")
     set(LIBOMP_ARCH i386)
   elseif(LIBOMP_NATIVE_ARCH STREQUAL "amd64")
-    set(LIBOMP_ARCH x86_64)
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      set(LIBOMP_ARCH x86_64_x32)
+    else()
+      set(LIBOMP_ARCH x86_64)
+    endif()
   elseif(LIBOMP_NATIVE_ARCH STREQUAL "x86_64")
-    set(LIBOMP_ARCH x86_64)
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      set(LIBOMP_ARCH x86_64_x32)
+    else()
+      set(LIBOMP_ARCH x86_64)
+    endif()
   elseif(LIBOMP_NATIVE_ARCH MATCHES "powerpc64le")
     set(LIBOMP_ARCH ppc64le)
   elseif(LIBOMP_NATIVE_ARCH MATCHES "powerpc64")
@@ -91,7 +99,7 @@
   endif()
 endif()
 
-libomp_check_variable(LIBOMP_ARCH 32e x86_64 32 i386 arm ppc ppc64 ppc64le aarch64 aarch64_32 aarch64_a64fx mic mips mips64 riscv64 loongarch64 ve s390x sparc sparcv9 wasm32)
+libomp_check_variable(LIBOMP_ARCH 32e x86_64_x32 x86_64 32 i386 arm ppc ppc64 ppc64le aarch64 aarch64_32 aarch64_a64fx mic mips mips64 riscv64 loongarch64 ve s390x sparc sparcv9 wasm32)
 
 set(LIBOMP_LIB_TYPE normal CACHE STRING
   "Performance,Profiling,Stubs library (normal/profile/stubs)")
@@ -168,6 +176,7 @@
 # Architecture
 set(IA32 FALSE)
 set(INTEL64 FALSE)
+set(X32 FALSE)
 set(ARM FALSE)
 set(AARCH64 FALSE)
 set(AARCH64_32 FALSE)
@@ -188,6 +197,8 @@
 set(SPARCV9 FALSE)
 if("${LIBOMP_ARCH}" STREQUAL "i386" OR "${LIBOMP_ARCH}" STREQUAL "32")    # IA-32 architecture
   set(IA32 TRUE)
+elseif("${LIBOMP_ARCH}" STREQUAL "x86_64_x32") # x32 ABI architecture
+  set(X32 TRUE)
 elseif("${LIBOMP_ARCH}" STREQUAL "x86_64" OR "${LIBOMP_ARCH}" STREQUAL "32e") # Intel(R) 64 architecture
   set(INTEL64 TRUE)
 elseif("${LIBOMP_ARCH}" STREQUAL "arm") # ARM architecture
--- a/openmp/runtime/cmake/LibompGetArchitecture.cmake	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/cmake/LibompGetArchitecture.cmake	2026-07-08 23:57:59.242711333 +0800
@@ -17,8 +17,12 @@
   set(detect_arch_src_txt "
     #if defined(__KNC__)
       #error ARCHITECTURE=mic
-    #elif defined(__amd64__) || defined(__amd64) || defined(__x86_64__) || defined(__x86_64) || defined(_M_X64) || defined(_M_AMD64)
+    #elif defined(__amd64__) || defined(__amd64) || defined(__x86_64__) || defined(__x86_64)
+    #if defined(__ILP32__)
+      #error ARCHITECTURE=x86_64_x32
+    #else
       #error ARCHITECTURE=x86_64
+    #endif
     #elif defined(__i386) || defined(__i386__) || defined(__IA32__) || defined(_M_I86) || defined(_M_IX86) || defined(__X86__) || defined(_X86_)
       #error ARCHITECTURE=i386
     #elif defined(__ARM_ARCH_7__) || defined(__ARM_ARCH_7R__) ||  defined(__ARM_ARCH_7A__) || defined(__ARM_ARCH_7M__)  || defined(__ARM_ARCH_7S__)
--- a/openmp/runtime/cmake/LibompMicroTests.cmake	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/cmake/LibompMicroTests.cmake	2026-07-08 23:57:59.242880418 +0800
@@ -75,6 +75,9 @@
   if(${IA32})
     libomp_append(libomp_test_touch_cflags -m32 LIBOMP_HAVE_M32_FLAG)
   endif()
+  if(${X32})
+    libomp_append(libomp_test_touch_cflags -mx32 LIBOMP_HAVE_M32_FLAG)
+  endif()
   libomp_append(libomp_test_touch_libs ${LIBOMP_OUTPUT_DIRECTORY}/${LIBOMP_LIB_FILE})
   libomp_append(libomp_test_touch_libs "${LIBOMP_HWLOC_LIBRARY}" LIBOMP_USE_HWLOC)
   if(APPLE)
@@ -193,6 +196,10 @@
       libomp_append(libomp_expected_library_deps libc.so.6)
       libomp_append(libomp_expected_library_deps ld-linux.so.2)
       libomp_append(libomp_expected_library_deps librt.so.1)
+    elseif(${X32})
+      libomp_append(libomp_expected_library_deps libc.so.6)
+      libomp_append(libomp_expected_library_deps ld-linux-x32.so.2)
+      libomp_append(libomp_expected_library_deps librt.so.1)
     elseif(${INTEL64})
       libomp_append(libomp_expected_library_deps libc.so.6)
       libomp_append(libomp_expected_library_deps ld-linux-x86-64.so.2)
--- a/openmp/runtime/cmake/config-ix.cmake	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/cmake/config-ix.cmake	2026-07-08 23:57:59.243038961 +0800
@@ -71,6 +71,7 @@
 check_cxx_compiler_flag(-ftls-model=initial-exec LIBOMP_HAVE_FTLS_MODEL_FLAG)
 libomp_check_architecture_flag(-mmic LIBOMP_HAVE_MMIC_FLAG)
 libomp_check_architecture_flag(-m32 LIBOMP_HAVE_M32_FLAG)
+libomp_check_architecture_flag(-mx32 LIBOMP_HAVE_M32_FLAG)
 if(WIN32)
   if(MSVC)
     # Check Windows MSVC style flags.
@@ -168,7 +169,7 @@
 endif()
 
 # Checking for x86-specific waitpkg and rtm attribute and intrinsics
-if (IA32 OR INTEL64)
+if (IA32 OR INTEL64 OR X32)
   check_include_file(immintrin.h LIBOMP_HAVE_IMMINTRIN_H)
   if (NOT LIBOMP_HAVE_IMMINTRIN_H)
     check_include_file(intrin.h LIBOMP_HAVE_INTRIN_H)
@@ -246,7 +247,7 @@
 endif()
 
 # Check if adaptive locks are available
-if((${IA32} OR ${INTEL64}) AND NOT MSVC)
+if((${IA32} OR ${INTEL64} OR ${X32}) AND NOT MSVC)
   set(LIBOMP_HAVE_ADAPTIVE_LOCKS TRUE)
 else()
   set(LIBOMP_HAVE_ADAPTIVE_LOCKS FALSE)
@@ -264,7 +265,7 @@
      { unsigned long long t = __builtin_readcyclecounter(); return 0; }"
      LIBOMP_HAVE___BUILTIN_READCYCLECOUNTER)
   if(NOT LIBOMP_HAVE___BUILTIN_READCYCLECOUNTER)
-    if(${IA32} OR ${INTEL64} OR ${MIC})
+    if(${IA32} OR ${INTEL64} OR ${X32} OR ${MIC})
       check_include_file(x86intrin.h LIBOMP_HAVE_X86INTRIN_H)
       libomp_append(CMAKE_REQUIRED_DEFINITIONS -DLIBOMP_HAVE_X86INTRIN_H LIBOMP_HAVE_X86INTRIN_H)
       check_c_source_compiles(
@@ -305,6 +306,7 @@
      ((LIBOMP_ARCH STREQUAL x86_64) OR
       (LIBOMP_ARCH STREQUAL i386) OR
 #      (LIBOMP_ARCH STREQUAL arm) OR
+      (LIBOMP_ARCH STREQUAL x86_64_x32) OR
       (LIBOMP_ARCH STREQUAL aarch64) OR
       (LIBOMP_ARCH STREQUAL aarch64_32) OR
       (LIBOMP_ARCH STREQUAL aarch64_a64fx) OR
@@ -349,7 +351,7 @@
 endif()
 
 # Check if ThreadSanitizer support is available
-if("${CMAKE_SYSTEM_NAME}" MATCHES "Linux" AND ${INTEL64})
+if("${CMAKE_SYSTEM_NAME}" MATCHES "Linux" AND (${INTEL64} OR ${X32}))
   set(LIBOMP_HAVE_TSAN_SUPPORT TRUE)
 else()
   set(LIBOMP_HAVE_TSAN_SUPPORT FALSE)
--- a/openmp/runtime/src/kmp.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp.h	2026-07-08 23:57:59.251453411 +0800
@@ -29,7 +29,7 @@
 #endif
 #define KMP_WEIGHTED_ITERATIONS_SUPPORTED                                      \
   (KMP_AFFINITY_SUPPORTED && KMP_STATIC_STEAL_ENABLED &&                       \
-   (KMP_ARCH_X86 || KMP_ARCH_X86_64))
+   (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32))
 
 #define TASK_CURRENT_NOT_QUEUED 0
 #define TASK_CURRENT_QUEUED 1
@@ -108,7 +108,7 @@
 #define KMP_HWLOC_ENABLED 0
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #include <xmmintrin.h>
 #endif
 
@@ -523,7 +523,7 @@
 #define KMP_FAST_REDUCTION_BARRIER 1
 
 #undef KMP_FAST_REDUCTION_CORE_DUO
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #define KMP_FAST_REDUCTION_CORE_DUO 1
 #endif
 
@@ -620,7 +620,7 @@
 
 typedef enum kmp_hw_core_type_t {
   KMP_HW_CORE_TYPE_UNKNOWN = 0x0,
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   KMP_HW_CORE_TYPE_ATOM = 0x20,
   KMP_HW_CORE_TYPE_CORE = 0x40,
   KMP_HW_MAX_NUM_CORE_TYPES = 3,
@@ -849,11 +849,11 @@
 
 enum affinity_top_method {
   affinity_top_method_all = 0, // try all (supported) methods, in order
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   affinity_top_method_apicid,
   affinity_top_method_x2apicid,
   affinity_top_method_x2apicid_1f,
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
   affinity_top_method_cpuinfo, // KMP_CPUINFO_FILE is usable on Windows* OS, too
 #if KMP_GROUP_AFFINITY
   affinity_top_method_group,
@@ -1282,7 +1282,7 @@
 #else
 #define KMP_BLOCKTIME(team, tid)                                               \
   (get__bt_set(team, tid) ? get__blocktime(team, tid) : __kmp_dflt_blocktime)
-#if KMP_OS_UNIX && (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if KMP_OS_UNIX && (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
 // HW TSC is used to reduce overhead (clock tick instead of nanosecond).
 extern kmp_uint64 __kmp_ticks_per_msec;
 extern kmp_uint64 __kmp_ticks_per_usec;
@@ -1335,7 +1335,7 @@
 /* Minimum number of threads before switch to TLS gtid (experimentally
    determined) */
 /* josh TODO: what about OS X* tuning? */
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #define KMP_TLS_GTID_MIN 5
 #else
 #define KMP_TLS_GTID_MIN INT_MAX
@@ -1403,7 +1403,7 @@
 #define KMP_NEXT_WAIT 512U /* susequent number of spin-tests */
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 typedef struct kmp_cpuid {
   kmp_uint32 eax;
   kmp_uint32 ebx;
@@ -3120,7 +3120,7 @@
 // t_inline_argv. Historically, we have supported at least 96 bytes. Using a
 // larger value for more space between the primary write/worker read section and
 // read/write by all section seems to buy more performance on EPCC PARALLEL.
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #define KMP_INLINE_ARGV_BYTES                                                  \
   (4 * CACHE_LINE -                                                            \
    ((3 * KMP_PTR_SKIP + 2 * sizeof(int) + 2 * sizeof(kmp_int8) +               \
@@ -3174,12 +3174,12 @@
   ompt_lw_taskteam_t *ompt_serialized_team_info;
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   kmp_int8 t_fp_control_saved;
   kmp_int8 t_pad2b;
   kmp_int16 t_x87_fpu_control_word; // FP control regs
   kmp_uint32 t_mxcsr;
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
   void *t_inline_argv[KMP_INLINE_ARGV_ENTRIES];
 
@@ -3205,7 +3205,7 @@
   omp_allocator_handle_t t_def_allocator; /* default allocator */
 
 // Read/write by workers as well
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
   // Using CACHE_LINE=64 reduces memory footprint, but causes a big perf
   // regression of epcc 'parallel' and 'barrier' on fxe256lin01. This extra
   // padding serves to fix the performance of epcc 'parallel' and 'barrier' when
@@ -3364,7 +3364,7 @@
                                          placement info */
 extern int __kmp_storage_map_verbose_specified;
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 extern kmp_cpuinfo_t __kmp_cpuinfo;
 static inline bool __kmp_is_hybrid_cpu() { return __kmp_cpuinfo.flags.hybrid; }
 #elif KMP_OS_DARWIN && KMP_ARCH_AARCH64
@@ -3529,11 +3529,11 @@
 #endif
 extern int __kmp_tls_gtid_min; /* #threads below which use sp search for gtid */
 extern int __kmp_foreign_tp; // If true, separate TP var for each foreign thread
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 extern int __kmp_inherit_fp_control; // copy fp creg(s) parent->workers at fork
 extern kmp_int16 __kmp_init_x87_fpu_control_word; // init thread's FP ctrl reg
 extern kmp_uint32 __kmp_init_mxcsr; /* init thread's mxscr */
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 // max_active_levels for nested parallelism enabled by default via
 // OMP_MAX_ACTIVE_LEVELS, OMP_NESTED, OMP_NUM_THREADS, and OMP_PROC_BIND
@@ -3901,7 +3901,7 @@
 extern void __kmp_expand_host_name(char *buffer, size_t size);
 extern void __kmp_expand_file_name(char *result, size_t rlen, char *pattern);
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || (KMP_OS_WINDOWS && (KMP_ARCH_AARCH64 || KMP_ARCH_ARM))
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || (KMP_OS_WINDOWS && (KMP_ARCH_AARCH64 || KMP_ARCH_ARM))
 extern void
 __kmp_initialize_system_tick(void); /* Initialize timer tick value */
 #endif
--- a/openmp/runtime/src/kmp_affinity.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_affinity.cpp	2026-07-08 23:57:59.243546751 +0800
@@ -174,7 +174,7 @@
   case KMP_HW_CORE_TYPE_UNKNOWN:
   case KMP_HW_MAX_NUM_CORE_TYPES:
     return "unknown";
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   case KMP_HW_CORE_TYPE_ATOM:
     return "Intel Atom(R) processor";
   case KMP_HW_CORE_TYPE_CORE:
@@ -1277,7 +1277,7 @@
     case KMP_HW_CORE_TYPE_UNKNOWN:
     case KMP_HW_MAX_NUM_CORE_TYPES:
       return 0;
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
     case KMP_HW_CORE_TYPE_ATOM:
       return 1;
     case KMP_HW_CORE_TYPE_CORE:
@@ -1866,7 +1866,7 @@
                                   0) == 0) {
         for (unsigned i = 0; i < nr_infos; ++i) {
           if (__kmp_str_match("CoreType", 8, infos[i].name)) {
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
             if (__kmp_str_match("IntelAtom", 9, infos[i].value)) {
               cpukinds[idx].core_type = KMP_HW_CORE_TYPE_ATOM;
               break;
@@ -2108,7 +2108,7 @@
 }
 #endif /* KMP_GROUP_AFFINITY */
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 template <kmp_uint32 LSB, kmp_uint32 MSB>
 static inline unsigned __kmp_extract_bits(kmp_uint32 v) {
@@ -3063,7 +3063,7 @@
   __kmp_free(proc_info);
   return true;
 }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #define osIdIndex 0
 #define threadIdIndex 1
@@ -3480,7 +3480,7 @@
           goto dup_field;
 #endif
         threadInfo[num_avail][osIdIndex] = val;
-#if KMP_OS_LINUX && !(KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if KMP_OS_LINUX && !(KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
         char path[256];
         KMP_SNPRINTF(
             path, sizeof(path),
@@ -4868,7 +4868,7 @@
     }
 #endif // KMP_HWLOC_ENABLED
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
     if (!success) {
       success = __kmp_affinity_create_x2apicid_map(&msg_id);
       if (!success && verbose && msg_id != kmp_i18n_null) {
@@ -4881,7 +4881,7 @@
         KMP_INFORM(AffInfoStr, env_var, __kmp_i18n_catgets(msg_id));
       }
     }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #if KMP_OS_LINUX || KMP_OS_AIX
     if (!success) {
@@ -4925,7 +4925,7 @@
   }
 #endif // KMP_HWLOC_ENABLED
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   else if (__kmp_affinity_top_method == affinity_top_method_x2apicid ||
            __kmp_affinity_top_method == affinity_top_method_x2apicid_1f) {
     success = __kmp_affinity_create_x2apicid_map(&msg_id);
@@ -4940,7 +4940,7 @@
       KMP_FATAL(MsgExiting, __kmp_i18n_catgets(msg_id));
     }
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
   else if (__kmp_affinity_top_method == affinity_top_method_cpuinfo) {
     int line = 0;
--- a/openmp/runtime/src/kmp_atomic.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_atomic.cpp	2026-07-08 23:57:59.244830191 +0800
@@ -865,7 +865,7 @@
 
 #endif // KMP_OS_WINDOWS && (KMP_ARCH_AARCH64 || KMP_ARCH_ARM)
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // ------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ====================================
@@ -941,7 +941,7 @@
   }
 // end of the second part of the workaround for C78287
 #endif // USE_CMPXCHG_FIX
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 // Routines for ATOMIC 4-byte operands addition and subtraction
 ATOMIC_FIXED_ADD(fixed4, add, kmp_int32, 32, +, 4i, 3,
@@ -1082,7 +1082,7 @@
   OP_CRITICAL(= *lhs OP, LCK_ID)                                               \
   }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // ------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ===================================
@@ -1105,7 +1105,7 @@
     OP_CRITICAL(= *lhs OP, LCK_ID) /* unaligned - use critical */              \
   }                                                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 ATOMIC_CMPX_L(fixed1, andl, char, 8, &&, 1i, 0,
               KMP_ARCH_X86) // __kmpc_atomic_fixed1_andl
@@ -1180,7 +1180,7 @@
   }                                                                            \
   }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // -------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ====================================
@@ -1209,7 +1209,7 @@
     }                                                                          \
   }                                                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 MIN_MAX_COMPXCHG(fixed1, max, char, 8, <, 1i, 0,
                  KMP_ARCH_X86) // __kmpc_atomic_fixed1_max
@@ -1235,12 +1235,12 @@
                  KMP_ARCH_X86) // __kmpc_atomic_float8_max
 MIN_MAX_COMPXCHG(float8, min, kmp_real64, 64, >, 8r, 7,
                  KMP_ARCH_X86) // __kmpc_atomic_float8_min
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 MIN_MAX_CRITICAL(float10, max, long double, <, 10r,
                  1) // __kmpc_atomic_float10_max
 MIN_MAX_CRITICAL(float10, min, long double, >, 10r,
                  1) // __kmpc_atomic_float10_min
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #if KMP_HAVE_QUAD
 MIN_MAX_CRITICAL(float16, max, QUAD_LEGACY, <, 16r,
                  1) // __kmpc_atomic_float16_max
@@ -1263,7 +1263,7 @@
   }
 
 // ------------------------------------------------------------------------
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // ------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ===================================
 #define ATOMIC_CMPX_EQV(TYPE_ID, OP_ID, TYPE, BITS, OP, LCK_ID, MASK,          \
@@ -1287,7 +1287,7 @@
     OP_CRITICAL(^= (TYPE) ~, LCK_ID) /* unaligned address - use critical */    \
   }                                                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 ATOMIC_CMPXCHG(fixed1, neqv, kmp_int8, 8, ^, 1i, 0,
                KMP_ARCH_X86) // __kmpc_atomic_fixed1_neqv
@@ -1319,7 +1319,7 @@
   }
 
 /* ------------------------------------------------------------------------- */
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // routines for long double type
 ATOMIC_CRITICAL(float10, add, long double, +, 10r,
                 1) // __kmpc_atomic_float10_add
@@ -1329,7 +1329,7 @@
                 1) // __kmpc_atomic_float10_mul
 ATOMIC_CRITICAL(float10, div, long double, /, 10r,
                 1) // __kmpc_atomic_float10_div
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #if KMP_HAVE_QUAD
 // routines for _Quad type
 ATOMIC_CRITICAL(float16, add, QUAD_LEGACY, +, 16r,
@@ -1375,7 +1375,7 @@
 ATOMIC_CRITICAL(cmplx8, sub, kmp_cmplx64, -, 16c, 1) // __kmpc_atomic_cmplx8_sub
 ATOMIC_CRITICAL(cmplx8, mul, kmp_cmplx64, *, 16c, 1) // __kmpc_atomic_cmplx8_mul
 ATOMIC_CRITICAL(cmplx8, div, kmp_cmplx64, /, 16c, 1) // __kmpc_atomic_cmplx8_div
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 ATOMIC_CRITICAL(cmplx10, add, kmp_cmplx80, +, 20c,
                 1) // __kmpc_atomic_cmplx10_add
 ATOMIC_CRITICAL(cmplx10, sub, kmp_cmplx80, -, 20c,
@@ -1384,7 +1384,7 @@
                 1) // __kmpc_atomic_cmplx10_mul
 ATOMIC_CRITICAL(cmplx10, div, kmp_cmplx80, /, 20c,
                 1) // __kmpc_atomic_cmplx10_div
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #if KMP_HAVE_QUAD
 ATOMIC_CRITICAL(cmplx16, add, CPLX128_LEG, +, 32c,
                 1) // __kmpc_atomic_cmplx16_add
@@ -1408,7 +1408,7 @@
 
 // OpenMP 4.0: x = expr binop x for non-commutative operations.
 // Supported only on IA-32 architecture and Intel(R) 64
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // ------------------------------------------------------------------------
 // Operation on *lhs, rhs bound by critical section
@@ -1613,7 +1613,7 @@
 #endif // KMP_ARCH_X86
 #endif // KMP_HAVE_QUAD
 
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // End of OpenMP 4.0: x = expr binop x for non-commutative operations.
 
 /* ------------------------------------------------------------------------ */
@@ -1644,7 +1644,7 @@
   }
 
 // -------------------------------------------------------------------------
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // -------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ====================================
 #define ATOMIC_CMPXCHG_MIX(TYPE_ID, TYPE, OP_ID, BITS, OP, RTYPE_ID, RTYPE,    \
@@ -1669,10 +1669,10 @@
                        LCK_ID) /* unaligned address - use critical */          \
   }                                                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 // -------------------------------------------------------------------------
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // -------------------------------------------------------------------------
 #define ATOMIC_CMPXCHG_REV_MIX(TYPE_ID, TYPE, OP_ID, BITS, OP, RTYPE_ID,       \
                                RTYPE, LCK_ID, MASK, GOMP_FLAG)                 \
@@ -1686,7 +1686,7 @@
   OP_GOMP_CRITICAL_REV(TYPE, OP, GOMP_FLAG)                                    \
   OP_CRITICAL_REV(TYPE, OP, LCK_ID)                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 // RHS=float8
 ATOMIC_CMPXCHG_MIX(fixed1, char, mul, 8, *, float8, kmp_real64, 1i, 0,
@@ -1803,7 +1803,7 @@
 ATOMIC_CMPXCHG_MIX(float8, kmp_real64, div, 64, /, fp, _Quad, 8r, 7,
                    KMP_ARCH_X86) // __kmpc_atomic_float8_div_fp
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 ATOMIC_CRITICAL_FP(float10, long double, add, +, fp, _Quad, 10r,
                    1) // __kmpc_atomic_float10_add_fp
 ATOMIC_CRITICAL_FP(float10, long double, sub, -, fp, _Quad, 10r,
@@ -1864,11 +1864,11 @@
                        1) // __kmpc_atomic_float10_sub_rev_fp
 ATOMIC_CRITICAL_REV_FP(float10, long double, div_rev, /, fp, _Quad, 10r,
                        1) // __kmpc_atomic_float10_div_rev_fp
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #endif // KMP_HAVE_QUAD
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // ------------------------------------------------------------------------
 // X86 or X86_64: no alignment problems ====================================
 #if USE_CMPXCHG_FIX
@@ -1903,7 +1903,7 @@
                        LCK_ID) /* unaligned address - use critical */          \
   }                                                                            \
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 ATOMIC_CMPXCHG_CMPLX(cmplx4, kmp_cmplx32, add, 64, +, cmplx8, kmp_cmplx64, 8c,
                      7, KMP_ARCH_X86) // __kmpc_atomic_cmplx4_add_cmplx8
@@ -2924,7 +2924,7 @@
 // binop x; v = x; }  for non-commutative operations.
 // Supported only on IA-32 architecture and Intel(R) 64
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // -------------------------------------------------------------------------
 // Operation on *lhs, rhs bound by critical section
 //     OP     - operator (it's supposed to contain an assignment)
@@ -3405,7 +3405,7 @@
 
 // End of OpenMP 4.0 Capture
 
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 #undef OP_CRITICAL
 
@@ -3464,7 +3464,7 @@
   if (
 #if KMP_ARCH_X86 && defined(KMP_GOMP_COMPAT)
       FALSE /* must use lock */
-#elif KMP_ARCH_X86 || KMP_ARCH_X86_64
+#elif KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
       TRUE /* no alignment problems */
 #else
       !((kmp_uintptr_t)lhs & 0x1) /* make sure address is 2-byte aligned */
@@ -3513,11 +3513,11 @@
   if (
 // FIXME: On IA-32 architecture, gcc uses cmpxchg only for 4-byte ints.
 // Gomp compatibility is broken if this routine is called for floats.
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
       TRUE /* no alignment problems */
 #else
       !((kmp_uintptr_t)lhs & 0x3) /* make sure address is 4-byte aligned */
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   ) {
     kmp_int32 old_value, new_value;
 
@@ -3563,7 +3563,7 @@
 
 #if KMP_ARCH_X86 && defined(KMP_GOMP_COMPAT)
       FALSE /* must use lock */
-#elif KMP_ARCH_X86 || KMP_ARCH_X86_64
+#elif KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
       TRUE /* no alignment problems */
 #else
       !((kmp_uintptr_t)lhs & 0x7) /* make sure address is 8-byte aligned */
@@ -3604,7 +3604,7 @@
       __kmp_release_atomic_lock(&__kmp_atomic_lock_8i, gtid);
   }
 }
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 void __kmpc_atomic_10(ident_t *id_ref, int gtid, void *lhs, void *rhs,
                       void (*f)(void *, void *, void *)) {
   KMP_DEBUG_ASSERT(__kmp_init_serial);
@@ -3625,7 +3625,7 @@
 #endif /* KMP_GOMP_COMPAT */
     __kmp_release_atomic_lock(&__kmp_atomic_lock_10r, gtid);
 }
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 void __kmpc_atomic_16(ident_t *id_ref, int gtid, void *lhs, void *rhs,
                       void (*f)(void *, void *, void *)) {
@@ -3647,7 +3647,7 @@
 #endif /* KMP_GOMP_COMPAT */
     __kmp_release_atomic_lock(&__kmp_atomic_lock_16c, gtid);
 }
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 void __kmpc_atomic_20(ident_t *id_ref, int gtid, void *lhs, void *rhs,
                       void (*f)(void *, void *, void *)) {
   KMP_DEBUG_ASSERT(__kmp_init_serial);
@@ -3668,7 +3668,7 @@
 #endif /* KMP_GOMP_COMPAT */
     __kmp_release_atomic_lock(&__kmp_atomic_lock_20c, gtid);
 }
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 void __kmpc_atomic_32(ident_t *id_ref, int gtid, void *lhs, void *rhs,
                       void (*f)(void *, void *, void *)) {
   KMP_DEBUG_ASSERT(__kmp_init_serial);
@@ -3705,7 +3705,7 @@
   __kmp_release_atomic_lock(&__kmp_atomic_lock, gtid);
 }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // OpenMP 5.1 compare and swap
 
@@ -3868,7 +3868,7 @@
 }
 
 // End OpenMP 5.1 compare + capture
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 /*!
 @}
--- a/openmp/runtime/src/kmp_atomic.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_atomic.h	2026-07-08 23:57:59.245627280 +0800
@@ -701,7 +701,7 @@
 
 // OpenMP 4.0: x = expr binop x for non-commutative operations.
 // Supported only on IA-32 architecture and Intel(R) 64
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 void __kmpc_atomic_fixed1_sub_rev(ident_t *id_ref, int gtid, char *lhs,
                                   char rhs);
@@ -801,7 +801,7 @@
 #endif
 #endif // KMP_HAVE_QUAD
 
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 // routines for mixed types
 
@@ -1417,7 +1417,7 @@
 
 // OpenMP 4.0: v = x = expr binop x; { v = x; x = expr binop x; } { x = expr
 // binop x; v = x; }  for non-commutative operations.
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 char __kmpc_atomic_fixed1_sub_cpt_rev(ident_t *id_ref, int gtid, char *lhs,
                                       char rhs, int flag);
 char __kmpc_atomic_fixed1_div_cpt_rev(ident_t *id_ref, int gtid, char *lhs,
@@ -1842,7 +1842,7 @@
 
 // End OpenMP 5.1 compare + capture
 
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 /* ------------------------------------------------------------------------ */
 
--- a/openmp/runtime/src/kmp_csupport.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_csupport.cpp	2026-07-08 23:57:59.246080417 +0800
@@ -711,13 +711,13 @@
 
     /* return to the parallel section */
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
     if (__kmp_inherit_fp_control && serial_team->t.t_fp_control_saved) {
       __kmp_clear_x87_fpu_status_word();
       __kmp_load_x87_fpu_control_word(&serial_team->t.t_x87_fpu_control_word);
       __kmp_load_mxcsr(&serial_team->t.t_mxcsr);
     }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
     __kmp_pop_current_task_from_thread(this_thr);
 #if OMPD_SUPPORT
@@ -1444,7 +1444,7 @@
 #define KMP_TSX_LOCK(seq) __kmp_user_lock_seq
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #define KMP_CPUINFO_RTM (__kmp_cpuinfo.flags.rtm)
 #else
 #define KMP_CPUINFO_RTM 0
@@ -3047,7 +3047,7 @@
   if ((__kmp_user_lock_kind == lk_tas) &&
       (sizeof(lck->tas.lk.poll) <= OMP_LOCK_T_SIZE)) {
 #if KMP_OS_LINUX &&                                                            \
-    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
+    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
 // "fast" path implemented to fix customer performance issue
 #if USE_ITT_BUILD
     __kmp_itt_lock_releasing((kmp_user_lock_p)user_lock);
@@ -3143,7 +3143,7 @@
       (sizeof(lck->tas.lk.poll) + sizeof(lck->tas.lk.depth_locked) <=
        OMP_NEST_LOCK_T_SIZE)) {
 #if KMP_OS_LINUX &&                                                            \
-    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
+    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
     // "fast" path implemented to fix customer performance issue
     kmp_tas_lock_t *tl = (kmp_tas_lock_t *)user_lock;
 #if USE_ITT_BUILD
--- a/openmp/runtime/src/kmp_global.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_global.cpp	2026-07-08 23:57:59.246690280 +0800
@@ -18,7 +18,7 @@
 
 kmp_key_t __kmp_gtid_threadprivate_key;
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 kmp_cpuinfo_t __kmp_cpuinfo = {0}; // Not initialized
 #endif
 
@@ -185,11 +185,11 @@
 #endif /* KMP_TDATA_GTID */
 int __kmp_tls_gtid_min = INT_MAX;
 int __kmp_foreign_tp = TRUE;
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 int __kmp_inherit_fp_control = TRUE;
 kmp_int16 __kmp_init_x87_fpu_control_word = 0;
 kmp_uint32 __kmp_init_mxcsr = 0;
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #ifdef USE_LOAD_BALANCE
 double __kmp_load_balance_interval = 1.0;
--- a/openmp/runtime/src/kmp_lock.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_lock.cpp	2026-07-08 23:57:59.246953033 +0800
@@ -2633,7 +2633,7 @@
 }
 
 // Time stamp counter
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #define __kmp_tsc() __kmp_hardware_timestamp()
 // Runtime's default backoff parameters
 kmp_backoff_t __kmp_spin_backoff_params = {1, 4096, 100};
--- a/openmp/runtime/src/kmp_lock.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_lock.h	2026-07-08 23:57:59.247488912 +0800
@@ -168,7 +168,7 @@
 #ifndef KMP_USE_FUTEX
 #define KMP_USE_FUTEX                                                          \
   (KMP_OS_LINUX &&                                                             \
-   (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64))
+   (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64))
 #endif
 #if KMP_USE_FUTEX
 
@@ -638,7 +638,7 @@
                                                    kmp_int32 gtid);
 
 #if KMP_OS_LINUX &&                                                            \
-    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
+    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
 
 #define __kmp_acquire_user_lock_with_checks(lck, gtid)                         \
   if (__kmp_user_lock_kind == lk_tas) {                                        \
@@ -683,7 +683,7 @@
                                                 kmp_int32 gtid);
 
 #if KMP_OS_LINUX &&                                                            \
-    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
+    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
 
 #include "kmp_i18n.h" /* AC: KMP_FATAL definition */
 extern int __kmp_env_consistency_check; /* AC: copy from kmp.h here */
@@ -747,7 +747,7 @@
 extern int (*__kmp_acquire_nested_user_lock_with_checks_)(kmp_user_lock_p lck,
                                                           kmp_int32 gtid);
 
-#if KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
 
 #define __kmp_acquire_nested_user_lock_with_checks(lck, gtid, depth)           \
   if (__kmp_user_lock_kind == lk_tas) {                                        \
@@ -796,7 +796,7 @@
 extern int (*__kmp_test_nested_user_lock_with_checks_)(kmp_user_lock_p lck,
                                                        kmp_int32 gtid);
 
-#if KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
 static inline int __kmp_test_nested_user_lock_with_checks(kmp_user_lock_p lck,
                                                           kmp_int32 gtid) {
   if (__kmp_user_lock_kind == lk_tas) {
@@ -1042,7 +1042,7 @@
 
 // Shortcuts
 #define KMP_USE_INLINED_TAS                                                    \
-  (KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM)) && 1
+  (KMP_OS_LINUX && (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM)) && 1
 #define KMP_USE_INLINED_FUTEX KMP_USE_FUTEX && 0
 
 // List of lock definitions; all nested locks are indirect locks.
--- a/openmp/runtime/src/kmp_os.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_os.h	2026-07-08 23:57:59.247769221 +0800
@@ -97,7 +97,7 @@
 
 /* Check for quad-precision extension. */
 #define KMP_HAVE_QUAD 0
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #if KMP_COMPILER_ICC || KMP_COMPILER_ICX
 /* _Quad is already defined for icc */
 #undef KMP_HAVE_QUAD
@@ -123,7 +123,7 @@
 #undef KMP_HAVE_QUAD
 #define KMP_HAVE_QUAD 1
 #endif
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #define KMP_USE_X87CONTROL 0
 #if KMP_OS_WINDOWS
@@ -302,10 +302,10 @@
 // All mwait code tests for UMWAIT first, so it should only fall back to ring3
 // MWAIT for KNL.
 #define KMP_HAVE_MWAIT                                                         \
-  ((KMP_ARCH_X86 || KMP_ARCH_X86_64) && (KMP_OS_LINUX || KMP_OS_WINDOWS) &&    \
+  ((KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && (KMP_OS_LINUX || KMP_OS_WINDOWS) &&    \
    !KMP_MIC2)
 #define KMP_HAVE_UMWAIT                                                        \
-  ((KMP_ARCH_X86 || KMP_ARCH_X86_64) && (KMP_OS_LINUX || KMP_OS_WINDOWS) &&    \
+  ((KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && (KMP_OS_LINUX || KMP_OS_WINDOWS) &&    \
    !KMP_MIC)
 
 #if KMP_OS_WINDOWS
@@ -719,7 +719,7 @@
 #define KMP_XCHG_REAL64(p, v) __kmp_xchg_real64((p), (v));
 #endif
 
-#elif (KMP_ASM_INTRINS && KMP_OS_UNIX) || !(KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#elif (KMP_ASM_INTRINS && KMP_OS_UNIX) || !(KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
 
 /* cast p to correct type so that proper intrinsic will be used */
 #define KMP_TEST_THEN_INC32(p)                                                 \
@@ -1066,7 +1066,7 @@
 #define KMP_MB() /* nothing to do */
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #if KMP_MIC
 // fence-style instructions do not exist, but lock; xaddl $0,(%rsp) can be used.
 // We shouldn't need it, though, since the ABI rules require that
@@ -1223,7 +1223,7 @@
 // dynamic user lock is turned on
 #if KMP_USE_DYNAMIC_LOCK
 // Visual studio can't handle the asm sections in this code
-#define KMP_USE_TSX (KMP_ARCH_X86 || KMP_ARCH_X86_64) && !KMP_COMPILER_MSVC
+#define KMP_USE_TSX (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && !KMP_COMPILER_MSVC
 #ifdef KMP_USE_ADAPTIVE_LOCKS
 #undef KMP_USE_ADAPTIVE_LOCKS
 #endif
@@ -1233,7 +1233,7 @@
 // Enable tick time conversion of ticks to seconds
 #if KMP_STATS_ENABLED
 #define KMP_HAVE_TICK_TIME                                                     \
-  (KMP_OS_LINUX && (KMP_MIC || KMP_ARCH_X86 || KMP_ARCH_X86_64))
+  (KMP_OS_LINUX && (KMP_MIC || KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32))
 #endif
 
 // Warning levels
--- a/openmp/runtime/src/kmp_platform.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_platform.h	2026-07-08 23:57:59.243216864 +0800
@@ -116,6 +116,7 @@
 /* ---------------------- Architecture recognition ------------------- */
 
 #define KMP_ARCH_X86 0
+#define KMP_ARCH_X86_X32 0
 #define KMP_ARCH_X86_64 0
 #define KMP_ARCH_AARCH64 0
 #define KMP_ARCH_AARCH64_32 0
@@ -150,8 +151,13 @@
 
 #if KMP_OS_UNIX
 #if defined __x86_64
+#if defined(__ILP32__)
+#undef KMP_ARCH_X86_X32
+#define KMP_ARCH_X86_X32 1
+#else
 #undef KMP_ARCH_X86_64
 #define KMP_ARCH_X86_64 1
+#endif
 #elif defined __i386
 #undef KMP_ARCH_X86
 #define KMP_ARCH_X86 1
@@ -279,19 +285,19 @@
 
 /* Specify 32 bit architectures here */
 #define KMP_32_BIT_ARCH                                                        \
-  (KMP_ARCH_X86 || KMP_ARCH_ARM || KMP_ARCH_MIPS || KMP_ARCH_WASM ||           \
+  (KMP_ARCH_X86 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_MIPS || KMP_ARCH_WASM || \
    KMP_ARCH_PPC || KMP_ARCH_AARCH64_32 || KMP_ARCH_SPARC32)
 
 // Platforms which support Intel(R) Many Integrated Core Architecture
 #define KMP_MIC_SUPPORTED                                                      \
-  ((KMP_ARCH_X86 || KMP_ARCH_X86_64) && (KMP_OS_LINUX || KMP_OS_WINDOWS))
+  ((KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && (KMP_OS_LINUX || KMP_OS_WINDOWS))
 
 // TODO: Fixme - This is clever, but really fugly
-#if (1 != KMP_ARCH_X86 + KMP_ARCH_X86_64 + KMP_ARCH_ARM + KMP_ARCH_PPC64 +     \
+#if (1 != KMP_ARCH_X86 + KMP_ARCH_X86_64 + KMP_ARCH_X86_X32 + KMP_ARCH_ARM + KMP_ARCH_PPC64 +     \
               KMP_ARCH_AARCH64 + KMP_ARCH_MIPS + KMP_ARCH_MIPS64 +             \
               KMP_ARCH_RISCV64 + KMP_ARCH_LOONGARCH64 + KMP_ARCH_VE +          \
               KMP_ARCH_S390X + KMP_ARCH_WASM + KMP_ARCH_PPC +                  \
-              KMP_ARCH_AARCH64_32 + KMP_ARCH_SPARC)
+              KMP_ARCH_AARCH64_32 + KMP_ARCH_SPARC)
 #error Unknown or unsupported architecture
 #endif
 
--- a/openmp/runtime/src/kmp_runtime.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_runtime.cpp	2026-07-08 23:57:59.248325309 +0800
@@ -1113,7 +1113,7 @@
   KMP_MB();
 }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // Propagate any changes to the floating point control registers out to the team
 // We try to avoid unnecessary writes to the relevant cache line in the team
 // structure, so we don't make changes unless they are needed.
@@ -1173,7 +1173,7 @@
 #else
 #define propagateFPControl(x) ((void)0)
 #define updateHWFPControl(x) ((void)0)
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 static void __kmp_alloc_argv_entries(int argc, kmp_team_t *team,
                                      int realloc); // forward declaration
@@ -1720,11 +1720,11 @@
 
 /* josh todo: hypothetical question: what do we do for OS X*? */
 #if KMP_OS_LINUX &&                                                            \
-    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
+    (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || KMP_ARCH_AARCH64)
   SimpleVLA<void *> args(argc);
 #else
   void **args = (void **)KMP_ALLOCA(argc * sizeof(void *));
-#endif /* KMP_OS_LINUX && ( KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_ARM || \
+#endif /* KMP_OS_LINUX && ( KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_ARM || \
           KMP_ARCH_AARCH64) */
 
   KA_TRACE(
@@ -4753,11 +4753,11 @@
   // TODO???: team->t.t_max_active_levels       = new_max_active_levels;
   team->t.t_sched.sched = new_icvs->sched.sched;
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   team->t.t_fp_control_saved = FALSE; /* not needed */
   team->t.t_x87_fpu_control_word = 0; /* not needed */
   team->t.t_mxcsr = 0; /* not needed */
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
   team->t.t_construct = 0;
 
@@ -6710,7 +6710,7 @@
     double dtime;
     long ltime;
   } time;
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   __kmp_initialize_system_tick();
 #endif
   __kmp_read_system_time(&time.dtime);
@@ -7522,13 +7522,13 @@
   KA_TRACE(10, ("__kmp_parallel_initialize: enter\n"));
   KMP_ASSERT(KMP_UBER_GTID(gtid));
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   // Save the FP control regs.
   // Worker threads will set theirs to these values at thread startup.
   __kmp_store_x87_fpu_control_word(&__kmp_init_x87_fpu_control_word);
   __kmp_store_mxcsr(&__kmp_init_mxcsr);
   __kmp_init_mxcsr &= KMP_X86_MXCSR_MASK;
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #if KMP_OS_UNIX
 #if KMP_HANDLE_SIGNALS
--- a/openmp/runtime/src/kmp_settings.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_settings.cpp	2026-07-08 23:57:59.249609011 +0800
@@ -789,7 +789,7 @@
 // -----------------------------------------------------------------------------
 // KMP_INHERIT_FP_CONTROL
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 static void __kmp_stg_parse_inherit_fp_control(char const *name,
                                                char const *value, void *data) {
@@ -803,7 +803,7 @@
 #endif /* KMP_DEBUG */
 } // __kmp_stg_print_inherit_fp_control
 
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 // Used for OMP_WAIT_POLICY
 static char const *blocktime_str = NULL;
@@ -2028,7 +2028,7 @@
   case KMP_HW_CORE_TYPE_UNKNOWN:
   case KMP_HW_MAX_NUM_CORE_TYPES:
     return "unknown";
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   case KMP_HW_CORE_TYPE_ATOM:
     return "intel_atom";
   case KMP_HW_CORE_TYPE_CORE:
@@ -3090,7 +3090,7 @@
           }
           scan++; // skip ':'
           SKIP_WS(scan);
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
           if (__kmp_match_str("intel_core", scan, &next)) {
             __kmp_affinity.core_attr_gran.core_type = KMP_HW_CORE_TYPE_CORE;
             __kmp_affinity.core_attr_gran.valid = 1;
@@ -3306,7 +3306,7 @@
     __kmp_affinity_top_method = affinity_top_method_hwloc;
   }
 #endif // KMP_HWLOC_ENABLED
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   else if (__kmp_str_match("cpuid_leaf31", 12, value) ||
            __kmp_str_match("cpuid 1f", 8, value) ||
            __kmp_str_match("cpuid 31", 8, value) ||
@@ -3364,7 +3364,7 @@
              __kmp_str_match("leaf4", 5, value)) {
     __kmp_affinity_top_method = affinity_top_method_apicid;
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
   else if (__kmp_str_match("/proc/cpuinfo", 2, value) ||
            __kmp_str_match("cpuinfo", 5, value)) {
     __kmp_affinity_top_method = affinity_top_method_cpuinfo;
@@ -3395,7 +3395,7 @@
     value = "all";
     break;
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   case affinity_top_method_x2apicid_1f:
     value = "x2APIC id leaf 0x1f";
     break;
@@ -3407,7 +3407,7 @@
   case affinity_top_method_apicid:
     value = "APIC id";
     break;
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #if KMP_HWLOC_ENABLED
   case affinity_top_method_hwloc:
@@ -5220,7 +5220,7 @@
       if (attr_ptr) {
         attr.clear();
         // save the attribute
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
         if (__kmp_str_match("intel_core", -1, attr_ptr + 1)) {
           attr.set_core_type(KMP_HW_CORE_TYPE_CORE);
         } else if (__kmp_str_match("intel_atom", -1, attr_ptr + 1)) {
@@ -5568,10 +5568,10 @@
      __kmp_stg_print_handle_signals, NULL, 0, 0},
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
     {"KMP_INHERIT_FP_CONTROL", __kmp_stg_parse_inherit_fp_control,
      __kmp_stg_print_inherit_fp_control, NULL, 0, 0},
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #ifdef KMP_GOMP_COMPAT
     {"GOMP_STACKSIZE", __kmp_stg_parse_stacksize, NULL, NULL, 0, 0},
--- a/openmp/runtime/src/kmp_stats.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_stats.cpp	2026-07-08 23:57:59.250910157 +0800
@@ -774,7 +774,7 @@
   fprintf(statsOut, "# Time of run: %s\n", &buffer[0]);
   if (gethostname(&hostName[0], sizeof(hostName)) == 0)
     fprintf(statsOut, "# Hostname: %s\n", &hostName[0]);
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   fprintf(statsOut, "# CPU:  %s\n", &__kmp_cpuinfo.name[0]);
   fprintf(statsOut, "# Family: %d, Model: %d, Stepping: %d\n",
           __kmp_cpuinfo.family, __kmp_cpuinfo.model, __kmp_cpuinfo.stepping);
--- a/openmp/runtime/src/kmp_stats_timing.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_stats_timing.cpp	2026-07-08 23:57:59.250805643 +0800
@@ -28,7 +28,7 @@
   // pretty bad assumption of 1GHz clock for MIC
   return 1 / ((double)1000 * 1.e6);
 }
-#elif KMP_ARCH_X86 || KMP_ARCH_X86_64
+#elif KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 #include <string.h>
 // Extract the value from the CPUID information
 double tsc_tick_count::tick_time() {
--- a/openmp/runtime/src/kmp_utility.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/kmp_utility.cpp	2026-07-08 23:57:59.251201514 +0800
@@ -18,7 +18,7 @@
 
 static const char *unknown = "unknown";
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 
 /* NOTE: If called before serial_initialize (i.e. from runtime_initialize), then
    the debugging package has not been initialized yet, and only "0" will print
@@ -217,7 +217,7 @@
   }
 }
 
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 void __kmp_expand_host_name(char *buffer, size_t size) {
   KMP_DEBUG_ASSERT(size >= sizeof(unknown));
--- a/openmp/runtime/src/z_Linux_asm.S	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/z_Linux_asm.S	2026-07-08 23:57:59.252164439 +0800
@@ -17,7 +17,7 @@
 
 #include "kmp_config.h"
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 .att_syntax
 
 # if defined(__ELF__) && defined(__CET__) && defined(__has_include)
--- a/openmp/runtime/src/z_Linux_util.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/z_Linux_util.cpp	2026-07-08 23:57:59.253078240 +0800
@@ -321,7 +321,7 @@
 
 #endif // KMP_USE_FUTEX
 
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_WASM) && (!KMP_ASM_INTRINS)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_WASM) && (!KMP_ASM_INTRINS)
 /* Only 32-bit "add-exchange" instruction on IA-32 architecture causes us to
    use compare_and_store for these routines */
 
@@ -437,7 +437,7 @@
   return old_value;
 }
 
-#endif /* (KMP_ARCH_X86 || KMP_ARCH_X86_64) && (! KMP_ASM_INTRINS) */
+#endif /* (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && (! KMP_ASM_INTRINS) */
 
 void __kmp_terminate_thread(int gtid) {
   int status;
@@ -566,12 +566,12 @@
   KMP_CHECK_SYSFAIL("pthread_setcancelstate", status);
 #endif
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   // Set FP control regs to be a copy of the parallel initialization thread's.
   __kmp_clear_x87_fpu_status_word();
   __kmp_load_x87_fpu_control_word(&__kmp_init_x87_fpu_control_word);
   __kmp_load_mxcsr(&__kmp_init_mxcsr);
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 #ifdef KMP_BLOCK_SIGNALS
   status = sigfillset(&new_set);
@@ -1948,11 +1948,11 @@
     return;
   }
 
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
   if (!__kmp_cpuinfo.initialized) {
     __kmp_query_cpuid(&__kmp_cpuinfo);
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
   __kmp_xproc = __kmp_get_xproc();
 
@@ -2084,7 +2084,7 @@
   return nsec;
 }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 /* Measure clock ticks per millisecond */
 void __kmp_initialize_system_tick() {
   kmp_uint64 now, nsec2, diff;
@@ -2736,7 +2736,7 @@
 
 #endif // USE_LOAD_BALANCE
 
-#if !(KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_MIC || KMP_ARCH_AARCH64 ||        \
+#if !(KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_MIC || KMP_ARCH_AARCH64 ||        \
       KMP_ARCH_PPC64 || KMP_ARCH_RISCV64 || KMP_ARCH_LOONGARCH64 ||            \
       KMP_ARCH_ARM || KMP_ARCH_VE || KMP_ARCH_S390X || KMP_ARCH_PPC_XCOFF ||   \
       KMP_ARCH_AARCH64_32)
--- a/openmp/runtime/src/z_Windows_NT-586_util.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/z_Windows_NT-586_util.cpp	2026-07-08 23:57:59.254185532 +0800
@@ -12,7 +12,7 @@
 
 #include "kmp.h"
 
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_AARCH64 || KMP_ARCH_ARM)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_AARCH64 || KMP_ARCH_ARM)
 /* Only 32-bit "add-exchange" instruction on IA-32 architecture causes us to
    use compare_and_store for these routines */
 
@@ -74,7 +74,7 @@
   return old_value;
 }
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 kmp_int8 __kmp_test_then_add8(volatile kmp_int8 *p, kmp_int8 d) {
   kmp_int64 old_value, new_value;
 
@@ -102,7 +102,7 @@
   return old_value;
 }
 #endif /* KMP_ARCH_X86 */
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 kmp_uint64 __kmp_test_then_or64(volatile kmp_uint64 *p, kmp_uint64 d) {
   kmp_uint64 old_value, new_value;
@@ -189,4 +189,4 @@
 }
 #endif
 
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_AARCH64 || KMP_ARCH_ARM */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 || KMP_ARCH_AARCH64 || KMP_ARCH_ARM */
--- a/openmp/runtime/src/z_Windows_NT_util.cpp	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/src/z_Windows_NT_util.cpp	2026-07-08 23:57:59.253792627 +0800
@@ -717,11 +717,11 @@
 #endif /* USE_ITT_BUILD */
   __kmp_initialize_system_tick();
 
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32)
   if (!__kmp_cpuinfo.initialized) {
     __kmp_query_cpuid(&__kmp_cpuinfo);
   }
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
 /* Set up minimum number of threads to switch to TLS gtid */
 #if KMP_OS_WINDOWS && !KMP_DYNAMIC_LIB
@@ -1009,12 +1009,12 @@
 
   __kmp_affinity_bind_init_mask(gtid);
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   // Set FP control regs to be a copy of the parallel initialization thread's.
   __kmp_clear_x87_fpu_status_word();
   __kmp_load_x87_fpu_control_word(&__kmp_init_x87_fpu_control_word);
   __kmp_load_mxcsr(&__kmp_init_mxcsr);
-#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 */
+#endif /* KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32 */
 
   if (__kmp_stkoffset > 0 && gtid > 0) {
     padding = KMP_ALLOCA(gtid * __kmp_stkoffset);
--- a/openmp/runtime/test/atomic/kmp_atomic_cas.c	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/test/atomic/kmp_atomic_cas.c	2026-07-08 23:57:59.254436128 +0800
@@ -35,7 +35,7 @@
 
 int main() {
   int ret = 0;
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   bool r;
   char c0 = 1;
   char c1 = 2;
@@ -182,6 +182,6 @@
     printf("passed\n");
 #else
   printf("Unsupported architecture, skipping test...\n");
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   return ret;
 }
--- a/openmp/runtime/test/atomic/kmp_atomic_cas_cpt.c	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/test/atomic/kmp_atomic_cas_cpt.c	2026-07-08 23:57:59.254312202 +0800
@@ -41,7 +41,7 @@
 
 int main() {
   int ret = 0;
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   bool r;
   char c0 = 1;
   char c1 = 2;
@@ -221,6 +221,6 @@
     printf("passed\n");
 #else
   printf("Unsupported architecture, skipping test...\n");
-#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64
+#endif // KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
   return ret;
 }
--- a/openmp/runtime/test/atomic/kmp_atomic_float10_max_min.c	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/test/atomic/kmp_atomic_float10_max_min.c	2026-07-08 23:57:59.254553466 +0800
@@ -28,7 +28,7 @@
 
 int main() {
   int ret = 0;
-#if (KMP_ARCH_X86 || KMP_ARCH_X86_64) && !defined(_MSC_VER)
+#if (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && !defined(_MSC_VER)
   long double s = 012.3456; // small
   long double e = 123.4567; // middle
   long double d = 234.5678; // big
@@ -158,6 +158,6 @@
     printf("passed\n");
 #else
   printf("Unsupported architecture, skipping test...\n");
-#endif // (KMP_ARCH_X86 || KMP_ARCH_X86_64) && !defined(_MSC_VER)
+#endif // (KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32) && !defined(_MSC_VER)
   return ret;
 }
--- a/openmp/runtime/test/ompt/callback.h	2026-06-16 06:33:53.000000000 +0800
+++ b/openmp/runtime/test/ompt/callback.h	2026-07-08 23:57:59.254696662 +0800
@@ -256,7 +256,7 @@
   define_ompt_label(id)                                                        \
       print_possible_return_addresses(get_ompt_label_address(id))
 
-#if KMP_ARCH_X86 || KMP_ARCH_X86_64
+#if KMP_ARCH_X86 || KMP_ARCH_X86_64 || KMP_ARCH_X86_X32
 // On X86 the NOP instruction is 1 byte long. In addition, the compiler inserts
 // a MOV instruction for non-void runtime functions which is 3 bytes long.
 #define print_possible_return_addresses(addr)                                  \
